python/ray/actor.py

```diff
@@ -254,9 +372,105 @@ def __ray_restore_from_checkpoint__(cls, pickled_checkpoint):
+                # The preceding task executed on this actor instance. Save the
+                # checkpoint.
+                print("Saving actor checkpoint. actor_counter = {}."
+                      .format(task_counter))
+                actor_key = b"Actor:" + worker.actor_id
+
+                try:
+                    checkpoint = worker.actors[
+                        worker.actor_id].__ray_save_checkpoint__()
+                    # Save the checkpoint in Redis. TODO(rkn): Checkpoints
+                    # should not be stored in Redis. Fix this.
+                    worker.redis_client.hset(
+                        actor_key,
+                        "checkpoint_{}".format(task_counter),
+                        checkpoint)
+                    # Remove the previous checkpoints if there is one.
+                    checkpoint_indices = get_checkpoint_indices(
+                        worker, worker.actor_id)
+                    for index in checkpoint_indices:
+                        if index < task_counter:
+                            worker.redis_client.hdel(
+                                actor_key, "checkpoint_{}".format(index))
+                # An exception was thrown. Save the error.
+                except Exception as error:
+                    # Checkpoint saves should not block execution on the actor,
+                    # so we still consider the task successful.
+                    pass
+            else:
+                # The preceding task has not yet executed on this actor
+                # instance. Try to resume from the most recent checkpoint.
+                checkpoint_index, checkpoint = get_actor_checkpoint(
+                    worker, worker.actor_id)
+                if checkpoint_index == task_counter:
+                    # The checkpoint matches ours. Resume the actor instance.
+                    try:
+                        actor = (worker.actor_class.
+                                 __ray_restore_from_checkpoint__(checkpoint))
+                        worker.actors[worker.actor_id] = actor
+                    # An exception was thrown. Save the error.
+                    except Exception as error:
+                        # We could not resume the checkpoint, so count the task
+                        # as failed.
+                        actor_checkpoint_failed = True
+                        pass
+                else:
+                    # We cannot resume a mismatching checkpoint, so count the
+                    # task as failed.
+                    actor_checkpoint_failed = True
+
+            # Fall back to lineage reconstruction if we were unable to load the
+            # checkpoint.
+            if actor_checkpoint_failed:
+                worker.local_scheduler_client.reconstruct_object(
+                    plasma_id.binary())
+                worker.local_scheduler_client.notify_unblocked()
+
+            return actor_checkpoint_failed, error
```

Comment:
If I run the following,

```python
import ray
ray.init()

@ray.remote(checkpoint_interval=1)
class Foo(object):
    def __init__(self):
        pass
    def method(self):
        pass
    def __ray_save__(self):
        raise Exception("failure")

f = Foo.remote()
```

Then the first call to 

```python
ray.get(f.method.remote())
```

fails with

```
Traceback (most recent call last):
  File "/Users/rkn/Workspace/ray/python/ray/worker.py", line 728, in _process_task
    *arguments)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 151, in actor_method_executor
    actor_checkpoint_failed, error = method(actor, *args)
  File "/Users/rkn/Workspace/ray/python/ray/actor.py", line 472, in __ray_checkpoint__
    return actor_checkpoint_failed, error
UnboundLocalError: local variable 'error' referenced before assignment
```

And the second call to 

```python
ray.get(f.method.remote())
```

hangs.

Not sure if we want to address that here.